addons/link_tracker/models/link_tracker.py

```diff
@@ -112,26 +113,60 @@ def _get_title_from_url(self, url):
 
         return title
 
-    @api.model
-    def create(self, vals):
-        create_vals = vals.copy()
+    @api.constrains('url', 'campaign_id', 'medium_id', 'source_id')
+    def _check_unicity(self):
+        """
+        Can not be implemented in SQL because we want to care about null values.
+        """
+        # build a query to fetch all needed link trackers at once
+        search_query = []
+        for tracker in self:
+            search_query = ['|'] + search_query + [
+                '&', '&', '&',
+                ('url', '=', tracker.url),
+                ('campaign_id', '=', tracker.campaign_id.id),
+                ('medium_id', '=', tracker.medium_id.id),
+                ('source_id', '=', tracker.source_id.id)
+            ]
+        all_link_trackers = self.search(search_query[1:])
```

Comment:
Meh, this [:1] is not really clean. Normally using expression.OR it should not be necessary anymore as first OR should be done with a void list + a real list (to check).

Reply:
Yes you are right, not needed with expression.OR :)